components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -170,228 +176,174 @@ void AssetDiscoveryManager::OnGetSolanaTokenRegistry(
     return;
   }
 
-  // Create unique list of addresses that matched eth_getLogs query
-  // and keep track of largest block discovered
-  base::flat_set<std::string> matching_contract_addresses;
-  uint256_t largest_block = 0;
-  for (const auto& log : logs) {
-    matching_contract_addresses.insert(base::ToLowerASCII(log.address));
-    if (log.block_number > largest_block) {
-      largest_block = log.block_number;
+  // Create a map of chain_id to a vector of contract addresses that have a
+  // balance greater than 0
+  std::map<std::string, std::vector<std::string>>
+      chain_id_to_contract_addresses_with_balance;
+
+  // Populate the map using the balances
+  for (size_t i = 0; i < balances.size(); i++) {
+    if (balances[i] && balances[i].value() !=
+                           "0x0000000000000000000000000000000000000000000000000"
+                           "000000000000000") {
+      chain_id_to_contract_addresses_with_balance[chain_id].push_back(
+          contract_addresses[i]);
     }
   }
 
-  std::vector<mojom::BlockchainTokenPtr> discovered_assets;
-  for (const auto& contract_address : matching_contract_addresses) {
-    if (!tokens_to_search.contains(contract_address)) {
-      continue;
-    }
-    mojom::BlockchainTokenPtr token =
-        std::move(tokens_to_search.at(contract_address));
-
-    if (!BraveWalletService::AddUserAsset(token.Clone(), prefs_)) {
-      continue;
-    }
-    discovered_assets.push_back(std::move(token));
-  }
+  std::move(barrier_callback).Run(chain_id_to_contract_addresses_with_balance);
+}
 
-  if (!triggered_by_accounts_added) {
-    // Update the last block discovered for this chain unless this
-    // was triggered by accounts being added
-    ScopedDictPrefUpdate update(prefs_,
-                                kBraveWalletNextAssetDiscoveryFromBlocks);
-    base::Value::Dict& next_asset_discovery_from_blocks = update.Get();
-    const auto path = base::StrCat({kEthereumPrefKey, ".", chain_id});
-    const std::string* current =
-        next_asset_discovery_from_blocks.FindStringByDottedPath(path);
-    uint256_t current_int = 0;
-    if (current) {
-      if (!HexValueToUint256(*current, &current_int)) {
-        CompleteDiscoverAssets(
-            chain_id, std::vector<mojom::BlockchainTokenPtr>(),
-            mojom::ProviderError::kInternalError,
-            l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR),
-            triggered_by_accounts_added);
-        return;
+void AssetDiscoveryManager::MergeDiscoveredEthAssets(
+    base::flat_map<std::string,
+                   base::flat_map<std::string, mojom::BlockchainTokenPtr>>
+        chain_id_to_contract_address_to_token,
+    bool triggered_by_accounts_added,
+    std::vector<std::map<std::string, std::vector<std::string>>>
+        discovered_assets) {
+  // Create a vector of BlockchainTokenPtrs to return
+  std::vector<mojom::BlockchainTokenPtr> discovered_tokens;
+  base::flat_set<std::string> seen_contract_addresses;
```

Comment:
Should it be moved inside `for (const auto& [chain_id, contract_addresses] : discovered_asset)` for loop?
I think we should check if the contract address has been seen per chain rather than a global one.

Reply:
Yes, and for the same reason, I think the user_asset_set should be created per chain ID as well.